File: internal/services/appconnection/app_connection.go
package appconnection

import (
	"context"
	"time"
)

type UpdateConnectionParams struct {
	Connected          bool
	SetConnected       bool
	LastConnectedAt    *time.Time
	SetLastConnectedAt bool
	Error              *string
	SetError           bool
}

type UpdateESPSerialConnectionParams struct {
	Connected          bool
	SetConnected       bool
	LastConnectedAt    *time.Time
	SetLastConnectedAt bool
	Error              *string
	SetError           bool
}

type UpdatePICSerialConnectionParams struct {
	Connected          bool
	SetConnected       bool
	LastConnectedAt    *time.Time
	SetLastConnectedAt bool
	Error              *string
	SetError           bool
}

type UpdateRFIDUSBConnectionParams struct {
	Connected          bool
	SetConnected       bool
	LastConnectedAt    *time.Time
	SetLastConnectedAt bool
	Error              *string
	SetError           bool
}

type Service interface {
	UpdateCloudConnection(ctx context.Context, params UpdateConnectionParams) error
	UpdateESPSerialConnection(ctx context.Context, params UpdateConnectionParams) error
	UpdatePICSerialConnection(ctx context.Context, params UpdateConnectionParams) error
	UpdateRFIDUSBConnection(ctx context.Context, params UpdateConnectionParams) error
}

type Repository interface {
	GetAppConnection(ctx context.Context) (AppConnection, error)
	UpdateCloudConnection(ctx context.Context, params UpdateConnectionParams) error
	UpdateESPSerialConnection(ctx context.Context, params UpdateConnectionParams) error
	UpdatePICSerialConnection(ctx context.Context, params UpdateConnectionParams) error
	UpdateRFIDUSBConnection(ctx context.Context, params UpdateConnectionParams) error
}

File: ui/src/types/app-connection.ts
export interface CloudConnection {
  connected: boolean
  lastConnectedAt?: string
  uptime: number
  error?: string
}

export interface ESPSerialConnection {
  connected: boolean
  lastConnectedAt?: string
  error?: string
}

export interface PICSerialConnection {
  connected: boolean
  lastConnectedAt?: string
  error?: string
}

export interface RFIDUSBConnection {
  connected: boolean
  lastConnectedAt?: string
  error?: string
}

export interface AppConnection {
  cloudConnection: CloudConnection
  espSerialConnection: ESPSerialConnection
  picSerialConnection: PICSerialConnection
  rfidUsbConnection: RFIDUSBConnection
}
